plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -168,23 +184,61 @@ with the following signature:
 
     idFInteger : all (f :: * -> *). f integer -> f integer
 
-where the type application is handled with 'TyAppRep'
+where the type application is handled with 'TyAppRep'. Note that the latter is defined as a
+@data family@:
+
+    data family TyAppRep (fun :: dom -> cod) (arg :: dom) :: cod
+
+We do that because a @data family@ can return a poly-kinded argument, which allows us to get an
+intrinsically kinded representation of PLC types. For example, an opaque @term@ whose type is an
+application of a type variable @f@ to a type variable @a@ is represented like this:
+
+    Opaque term (TyAppRep (TyVarRep ('TyNameRep "f" 0)) (TyVarRep ('TyNameRep "a" 1 :: TyNameRep *)))
+
+Note the @TyNameRep *@ kind annotation. It tells us three things:
+
+1. a type-level name has a kind associated with it. The reason for that is that we use names in
+binders (for example, in the universal quantifier) and as variables and kinds are important in
+both these cases due to us having an intrinsically kinded representation of types, so it's
+convenient to annotate type-level names with kinds. Another reason is that we do not attempt to do
+any kind of static analysis on reflected type variables and associating kinds with them allows us
+to catch errors like "two variables with equal names and uniques, but different kinds" earlier
+2. the kind is not stored as an argument to the @TyNameRep@ constructor. Instead it's stored as
+an index of the data type. The point of this is that if we stored the kind as an argument, we'd
+have to provide it manually, but since the representation is intrinsically kinded there's no point
+in doing so as GHC can infer all the kinds itself
```

Comment:
```suggestion
in doing so as GHC can usually infer all the kinds itself
```

Reply:
I like the current wording. "GHC can infer all the kinds apart from cases where they're inherently ambiguous". Adding "usually" suggests that GHC occasionally fails to infer kinds in cases where they are not inherently ambiguous and I'm currently not aware of any such cases.

Reply:
Does that sound sensible?